### ISSUE TITLE
World Materials Seemingly Missing and Scrambled

### ISSUE THREAD

**Issue by Enverex** [Neos-Metaverse/NeosPublic#659]
This is going to be vague because in reality I only have the before and after situations I can show.

I have a world that has been fine for some time (Proteus Base). I ran all the Neos optimisation tools on it when it was first imported. No materials were deduplicated (none were duplicated so that was expected behaviour). I've worked on it many times since and it's been fine.

A few days ago I saw a picture of the world and it looked wrong. I logged in to find a lot of the objects in the world looked wrong. Specifically they appear to have had their materials swapped for another material in the world - Wall_Atlas_03_Flat and Wall_Atlas_04_Flat have all been swapped to Wall_Atlas_02_Flat. Those two missing/swapped materials are now also not present in the Assets folder in the hierarchy so they are just completely gone.

What's more odd is that I don't think the world has been saved between it working and it being scrambled. The world was fine last time I saved it.

Alex has also reported the same thing happening, except in his instance all the materials in the world were replaced with the same material.

**Comment by shiftyscales** [comment]
Hmm- in Alex's case from what I remember it wasn't all using the same material, but rather the materials were literally shuffled.

There is a mechanism that exists to replace all of one instance of a material with another. With the material tip set to scene global mode, it will replace all instances of a material in a world with the material you apply.

You can even shoot a material orb with it. That would fit more in line with what you are describing.

**Comment by Frooxius** [comment]
Hmm this is strange. Can you provide link to this world? Have you saved the world at all since this has happened? I could try loading it in and checking if something gets messed up during the loading process, but if you've saved it since, then the corrupted state was already written in.

Do you know which Neos build did it last work on correctly?

**Comment by Enverex** [comment]
I couldn't say when it last worked correctly time wise as I'd not been in there for a while. The world is public published so you can find it in the Worlds list (Featured too I believe) named "Proteus Base". I don't believe I've saved the world since this happened either, at least I don't remember doing so (I was going to try and fix it today but gave up mid-way and didn't save).

**Comment by Enverex** [comment]
So this happened again today, turning this:
![image](https://user-images.githubusercontent.com/958670/87883529-eb50b580-c9ff-11ea-8b22-d0d22a5100f3.png)

Into this:
![image](https://user-images.githubusercontent.com/958670/87883530-f0156980-c9ff-11ea-8fcb-2cc68c38328a.png)

Example broken material showing the new generic "WorldRoot - Assets" name rather than the original material name:
![image](https://user-images.githubusercontent.com/958670/87883560-2bb03380-ca00-11ea-8aaa-11ff1ed60eb4.png)


To clarify, at no point did I run merge materials and "Cleanup Unused Assets" was off the entire time this time.

So I think what is happening is this: When I save the hierarchy (titled WorldRoot containing the whole scene) to my inventory (so I can take it out of local space into cloud space) it breaks all the naming and renames everything to be generic names (in this case "WorldRoot - Assets" and not it's original material name). I then import that into the new cloud world and save it, in doing so, it seems to assume that a lot of those materials are in fact the same (for some reason, despite not being, except for the name) and in turn deduplicates them automatically on save.

Logs show deduplication taking place:
```19:20:01.227 ( 62 FPS)    World Optimized! Deduplicated Materials: 26, Deduplicated Static Providers: 6, Cleaned Up Assets: 92```

Let me know if you need anything else.

**Comment by shiftyscales** [comment]
Please provide the full logs for this issue happening. A single line/snippet doesn't give the full context for what all might have happened. Ideally if you can replicate this case in isolation by replicating the steps you suggested, that would be a huge lead if it works out, @Enverex.

**Comment by Enverex** [comment]
Full log of that session:
[JASPER - 2020.7.17.1288 - 2020-07-19 14_04_32.log](https://github.com/Frooxius/NeosPublic/files/4944510/JASPER.-.2020.7.17.1288.-.2020-07-19.14_04_32.log)

The workflow was: Import scene via Unity SDK (that won't be in this log, but that part was still fine -> (This is where the log starts) Work on the world in Neos as usual -> Save and close -> Reload local world, everything is still fine ->Save world hierarchy to inventory (this appears to mangle all the material names) -> Start new cloud world -> Import saved hierarchy -> Save and close -> Reopen new cloud session to test.

**Comment by Enverex** [comment]
Shifty managed to trigger this today. They were taking a copy of one of my worlds to check for another issue, as they saved a copy of the world, the materials changed, from this:
![image](https://user-images.githubusercontent.com/958670/91231126-a6b5e980-e724-11ea-9705-3e18743b278f.png)

To this:
![image](https://user-images.githubusercontent.com/958670/91231130-a9b0da00-e724-11ea-9115-c8306594e0fb.png)

If Shifty opens the copy they made, it mirrors the broken image. If I reload the original (non-copy) of the world, it looks correct.

**Comment by Enverex** [comment]
I think I tracked down what happened here. It seems to happen in two different scenarios:

1. Content hasn't finished uploading.

2. Content hasn't finished downloading.

If you then save the world, or export it into your inventory via the hierarchy while it's in this state, it appears to end up merging any of the "not yet loaded" content. Maybe because the materials technically don't have textures at that point because they've not uploaded/been assigned yet, I couldn't say. But it seems to result in the material duplicator merging those materials together resulting in this issue. This could be a red herring, but it seemed to tie in with what happened in each instance so far.

**Comment by chemicalcrux** [comment]
I've been consistently experiencing this in a reasonably large scene I'm working on. I have not yet worked out if it always triggers *after* saving.

I'll see if I can come up with a simple reproduction. I don't think it's caused by name collisions, because it just happened to a completely uniquely named mesh.

**Comment by shiftyscales** [comment]
> 
> 
> I've been consistently experiencing this in a reasonably large scene I'm working on. I have not yet worked out if it always triggers _after_ saving.
> 
> I'll see if I can come up with a simple reproduction. I don't think it's caused by name collisions, because it just happened to a completely uniquely named mesh.

I was able to replicate this issue previously by trying to save a copy of a world while it was still in the middle of loading in.

E.g. I opened a copy of Viridian Island, and while it was still loading I saved a copy of my own. In the copy I saved, a bunch of the materials were scrambled.

**Comment by chemicalcrux** [comment]
Hmm, that could be a separate cause, then -- a long time passed between the import and the first save, I never exited or entered the world, and nobody else was connected.

**Comment by Frooxius** [comment]
I've added a mechanism that will block any of the asset optimizations running on imports in 2020.12.18.1284. This is automatic, you don't need to do anything youself to make it work. Please reopen if this issue still occurs, due to its randomness it's a bit hard to test reliably.

**Comment by Khosumi** [comment]
I've had this issue happen to me just now, all of my materials are wrong in a world I did not save previously before logging in.

**Comment by H3BO3** [comment]
@Khosumi can you elaborate the sequence of events more? I'm not sure how you had a world you didn't save before starting neos

**Comment by Khosumi** [comment]
I mean I logged off in my world, logged back in at a later time and the mats were wrong.

**Comment by H3BO3** [comment]
@Khosumi And when was the last time you saved in any way? Is this a homeworld?

**Comment by Khosumi** [comment]
Yes, this is my favorited home world, I last saved about yesterday. Today, I logged in and out a couple times but I have made no changes to the mats myself.

**Comment by H3BO3** [comment]
@Khosumi And how did you exit the game those times?

**Comment by Khosumi** [comment]
Exit and discard

**Comment by H3BO3** [comment]
I see, thanks for the info! Are you able to submit the log file from when you started the game today and found the scrambled materials?

**Comment by Khosumi** [comment]
[DESKTOP-H0D0TE7 - 2021.7.1.498 - 2021-07-02 00_21_54.log](https://github.com/Neos-Metaverse/NeosPublic/files/6752155/DESKTOP-H0D0TE7.-.2021.7.1.498.-.2021-07-02.00_21_54.log)
I am pretty sure this is the log in question, as it looks like it didn't completely exit properly.

**Comment by Khosumi** [comment]
[DESKTOP-H0D0TE7 - 2021.7.1.498 - 2021-07-02 00_17_31.log](https://github.com/Neos-Metaverse/NeosPublic/files/6752201/DESKTOP-H0D0TE7.-.2021.7.1.498.-.2021-07-02.00_17_31.log)
This log shows the world being saved with the materials being "Optimized"

**Comment by H3BO3** [comment]
I believe the log above contains the necessary info.

https://discord.com/channels/402159838827905024/439422161657135106/860385226542022686 More context here.
